core/tx_list.go

```diff
@@ -529,21 +556,36 @@ func (l *txPricedList) Underpriced(tx *types.Transaction) bool {
 // Note local transaction won't be considered for eviction.
 func (l *txPricedList) Discard(slots int, force bool) (types.Transactions, bool) {
 	drop := make(types.Transactions, 0, slots) // Remote underpriced transactions to drop
-	for len(*l.remotes) > 0 && slots > 0 {
-		// Discard stale transactions if found during cleanup
-		tx := heap.Pop(l.remotes).(*types.Transaction)
-		if l.all.GetRemote(tx.Hash()) == nil { // Removed or migrated
-			l.stales--
-			continue
+	for slots > 0 {
+		if len(l.urgent.list)*floatingRatio > len(l.floating.list)*urgentRatio || floatingRatio == 0 {
+			// Discard stale transactions if found during cleanup
+			tx := heap.Pop(&l.urgent).(*types.Transaction)
+			if l.all.GetRemote(tx.Hash()) == nil { // Removed or migrated
+				l.stales--
+				continue
+			}
+			// Non stale transaction found, move to floating heap
+			heap.Push(&l.floating, tx)
+		} else {
+			if len(l.floating.list) == 0 {
+				// Stop if both heaps are empty
+				break
+			}
+			// Discard stale transactions if found during cleanup
+			tx := heap.Pop(&l.floating).(*types.Transaction)
+			if l.all.GetRemote(tx.Hash()) == nil { // Removed or migrated
+				l.stales--
+				continue
+			}
+			// Non stale transaction found, discard it
+			drop = append(drop, tx)
+			slots -= numSlots(tx)
 		}
-		// Non stale transaction found, discard it
-		drop = append(drop, tx)
-		slots -= numSlots(tx)
 	}
 	// If we still can't make enough room for the new transaction
 	if slots > 0 && !force {
 		for _, tx := range drop {
-			heap.Push(l.remotes, tx)
+			heap.Push(&l.urgent, tx)
```

Comment:
This is incorrect. The transaction most probably originated from the floating heap, not the urgent heap. Furthermore, we have done certain urgent -> floating heap movements, which should also be reverted in case we can't discard the required number of slots.

Reply:
With reverting urgent->floating movements you are theoretically right, I'll think about a clean and simple way to handle this.Putting everything in urgent is good though, we can always add everything to urgent first and balance the queues out before discarding anything.